api/METRICS.md

```diff
@@ -86,21 +86,186 @@ group_max_from_summary(summary)
+
+  * `"<metric>"`: <br>
+    goal is then to maximize _metric_(\*) subject to constraints
+
+### fairlearn.reductions
+
+We support two algorithms: `ExponentiatedGradient` and `GridSearch`. They both represent `constraints` and `objective` via objects of type `Moment`.
+
+**Status quo:**
+
+* In both cases, `objective` is automatically inferred from the provided `constraints`.
+
+* `ExponentiatedGradient(estimator, constraints, eps=`&epsilon;`)` considers constraints that are specified jointly by the provided `Moment` object and the numeric value &epsilon; as follows:
+
+  * `DemographicParity()`: for all _a_, <br>
+    |_selection_rate_(_a_) - _selection_rate_(\*)| &le; &epsilon;.
+
+  * `DemographicParity(ratio=`_r_`)`: for all _a_, <br>
+    _r_ &sdot; _selection_rate_(_a_) - _selection_rate_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _selection_rate_(\*) - _selection_rate_(_a_) &le; &epsilon;.
+
+  * `TruePositiveRateDifference()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;.
+
+  * `TruePositiveRateDifference(ratio=`_r_`)`: analogous
+
+  * `EqualizedOdds()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;, <br>
+    |_false_positive_rate_(_a_) - _false_positive_rate_(\*)| &le; &epsilon;.
+
+  * `EqualizedOdds(ratio=`_r_`)`: analogous
+
+  * `ErrorRateRatio()`: for all _a_, <br>
+    |_error_rate_(_a_) - _error_rate_(\*)| &le; &epsilon;.
+
+  * `ErrorRateDifference(ratio=`_r_`)`: analogous
+
+  * all of the above constraints are descendants of `ConditionalSelectionRate`
+
+  * the `objective` for all of the above constraints is `ErrorRate()`
+
+* `GridSearch(estimator, constraints)` considers constraints represented by the provided `Moment` object. The behavior of the `GridSearch` algorithm does not depend on the value of the right-hand side of constraints, so it is not provided to the constructor. In addition to the `Moment` objects above, `GridSearch` also supports the following:
+
+  * `GroupLossMoment(<loss>)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    * where `<loss>` is a _loss evaluation_ object; it supports an interface that takes `y_true` and `y_pred` as the input and returns the vector of losses evaluated on individual examples
+
+    * the `objective` for `GroupLossMoment(<loss>)` is `AverageLossMoment(<loss>)`
+
+  * both `GroupLossMoment` and `AverageLossMoment` are descendants of `ConditionalLossMoment`
+
+**Proposal.** The proposal introduces some breaking changes:
+
+* `constraints`:
+
+  * `<Metric>Parity(difference_bound=`&epsilon;`)`: for all _a_, <br>
+    |_metric_(_a_) - _metric_(\*)| &le; &epsilon;.
+
+  * `<Metric>Parity(ratio_bound=`_r_`, ratio_bound_slack=`&epsilon;`)`: for all _a_, <br>
+    _r_ &sdot; _metric_(_a_) - _metric_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _metric_(\*) - _metric_(_a_) &le; &epsilon;.
+
+  * `DemographicParity` and `EqualizedOdds` have the same calling convention as `<Metric>Parity`

```

Comment:
Does this mean that we should be able to run direct comparisons between things in `metrics` and things in `moments` to make sure they agree that they're calculating the same thing?

Reply:
Not quite, but almost... there's a small discrepancy between metrics and moments:
* `<metric>_difference`  and `<metric>_ratio` look at the difference / ratio between the max and the min
* `<Metric>Parity` is looking at the max difference/ratio between any group and the overall metric

I think that the final solution would be to enable a flag `relative_to_overall` in `<metric>_difference` and `<metric>_ratio` to get the same functionality as in the moments when the flag equals `True` and the current functionality when the flag equals `False`. But I'm not sure we need to put it in this proposal (but definitely will put it in the user guide).

Reply:
Great catch, Richard. I think this definitely needs to be documented well.